drv/lpc55-gpio-api/src/lib.rs

```diff
@@ -205,6 +205,9 @@ pub enum Value {
 #[repr(u32)]
 pub enum GpioError {
     BadArg = 2,
+
+    #[idol(server_death)]
+    ServerRestarted,
```

Comment:
I'd vote for this to manually implement `IHaveConsideredServerDeathWithThisErrorType` or become idempotent (probably the latter?); GPIO operations really shouldn't fail.

Reply:
In fact, this error appears to never be constructed!

Reply:
I think the problem with `idempotent` on GPIO calls is `toggle()`, which does not seem like something you want an idol client to retry some unknown number of times. The STM32 GPIO idol interface marks all the operations except `toggle` as idempotent. I can do the same for `lpc55-pins` but will still need this error for `toggle`.

Reply:
`toggle` appears to never actually construct an error:
```rust
    fn toggle(
        &mut self,
        _: &RecvMessage,
        pin: Pin,
    ) -> Result<(), RequestError<GpioError>> {
        let (port, pin) = gpio_port_pin_validate(pin);

        self.gpio.not[port].write(|w| unsafe { w.notp().bits(1 << pin) });
        Ok(())
    }
```
(which is why I could delete the members of `GpioError` and stuff still compiled)
 
I think this means we could return `Simple("()")`?

Reply:
Ah right, I was thinking of `err: ServerDeath` instead.

This causes Idolatry to generate a function which returns `Result<(), idol_runtime::ServerDeath>`, where the latter is an empty marker:

```diff
diff --git a/drv/lpc55-gpio/src/main.rs b/drv/lpc55-gpio/src/main.rs
index a3c58ca9..0654b9fd 100644
--- a/drv/lpc55-gpio/src/main.rs
+++ b/drv/lpc55-gpio/src/main.rs
@@ -111,7 +111,7 @@ impl idl::InOrderPinsImpl for ServerImpl<'_> {
         &mut self,
         _: &RecvMessage,
         pin: Pin,
-    ) -> Result<(), RequestError<GpioError>> {
+    ) -> Result<(), RequestError<core::convert::Infallible>> {
         let (port, pin) = gpio_port_pin_validate(pin);
 
         self.gpio.not[port].write(|w| unsafe { w.notp().bits(1 << pin) });
diff --git a/idl/lpc55-pins.idol b/idl/lpc55-pins.idol
index 36b89e06..cdd3c646 100644
--- a/idl/lpc55-pins.idol
+++ b/idl/lpc55-pins.idol
@@ -46,7 +46,7 @@ Interface(
 		},
 		reply: Result(
 			ok: "()",
-			err: CLike("GpioError"),
+			err: ServerDeath,
 		)
 	),
 
diff --git a/task/hiffy/src/lpc55.rs b/task/hiffy/src/lpc55.rs
index 63c098fc..9164077e 100644
--- a/task/hiffy/src/lpc55.rs
+++ b/task/hiffy/src/lpc55.rs
@@ -273,7 +273,7 @@ fn gpio_toggle(
 
     match gpio.toggle(pin) {
         Ok(_) => Ok(0),
-        Err(err) => Err(Failure::FunctionError(err.into())),
+        Err(_) => panic!(), // only error possible is server death
     }
 }
```

Reply:
Good call; I forgot about `ServerDeath`. I made the change to lpc55 gpios in 4ea0f07 and stm32 gpios in 69a720d.